### PR TITLE
CORS-3748: IBMCloud: Setup IBM Cloud CAPI InfraReady step

### PR DESCRIPTION
**Pull request by cjschaef**
Configure InfraReady to create DNS Records for Infrastructure Load Balancers.

Related: https://issues.redhat.com/browse/CORS-3748

### CODE REVIEW

**Comment by cjschaef** [comment]
Load Balancer provisioning/reconciliation is not merged into IBM Cloud CAPI, which this process is dependent upon.
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/2023

**Comment by cjschaef** [comment]
/retitle CORS-3748: IBMCloud: Setup IBM Cloud CAPI InfraReady step

**Comment by cjschaef** [comment]
Test failure is a known issue:
```

: [Jira:"Networking / router"] monitor test service-type-load-balancer-availability setup expand_less | 45s
-- | --
{  failed during setup error waiting for replicaset: failed to create RC "service-test": Internal error occurred: resource quota evaluation timed out}
```
https://issues.redhat.com/browse/OCPBUGS-42869

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 1]
```diff
@@ -137,3 +167,68 @@ func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionIn
...
+			return fmt.Errorf("load balancer %s is not ready, infrastructure not ready: %s", lbID, lb.State)
+		}
+		// Lookup Load Balancer details to use during DNS Record creation.
+		lbDetails, err := client.GetLoadBalancer(context.TODO(), lbID)
+		if err != nil {
+			return fmt.Errorf("failed retrieving load balancer for dns record creation: %w", err)
+		} else if lbDetails == nil {
+			return fmt.Errorf("failed to find load balancer for dns record creation by id: %s", lbID)
+		}
+		err = metadata.CreateDNSRecord(context.TODO(), in.InstallConfig.Config.ClusterDomain(), lbDetails)
```
Comment (r4f4) [inline pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go:218]:
```suggestion
		err = metadata.CreateDNSRecord(ctx, in.InstallConfig.Config.ClusterDomain(), lbDetails)
```

**Comment by r4f4** [comment]
@cjschaef Bear in mind that `e2e-ibmcloud-ovn` is still terraform-based and you need to trigger `altinfra-e2e-ibmcloud-capi-ovn` for the capi-based installation.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go (hunk 1 above)
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: updated.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 2]
```diff
@@ -137,3 +167,68 @@ func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionIn
...
+		return fmt.Errorf("load balancers missing from ibmcloudCluster.Status in InfraReady")
+	}
+
+	for lbID, lb := range ibmcloudCluster.Status.Network.LoadBalancers {
+		// Verify that the Load Balancer is ready (active).
+		if lb.State != capibmcloud.VPCLoadBalancerStateActive {
+			return fmt.Errorf("load balancer %s is not ready, infrastructure not ready: %s", lbID, lb.State)
+		}
+		// Lookup Load Balancer details to use during DNS Record creation.
+		lbDetails, err := client.GetLoadBalancer(context.TODO(), lbID)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: updated.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 3]
```diff
@@ -137,3 +167,68 @@ func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionIn
...
+	}
+	logrus.Debugf("InfraReady: ibmcloudCluster = %+v", ibmcloudCluster)
+	logrus.Debugf("InfraReady: ibmcloudCluster.Status = %+v", ibmcloudCluster.Status)
+	if ibmcloudCluster.Status.Network == nil || ibmcloudCluster.Status.Network.VPC == nil || ibmcloudCluster.Status.Network.VPC.ID == "" {
+		return fmt.Errorf("vpc missing from ibmcloudCluster.Status in InfraReady")
+	}
+	logrus.Debugf("InfraReady: ibmcloudCluster.Status.Network.VPC.ID = %s", ibmcloudCluster.Status.Network.VPC.ID)
+
+	// Collect the Load Balancer's and their hostnames to use for DNS Record creation.
+	if ibmcloudCluster.Status.Network.LoadBalancers == nil || len(ibmcloudCluster.Status.Network.LoadBalancers) == 0 {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: updated.

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 4]
```diff
@@ -217,6 +220,63 @@ func (c *Client) CreateCOSObject(ctx context.Context, sourceData []byte, fileNam
...
+
+	dnsService, err := c.getDNSServicesAPI()
+	if err != nil {
+		return fmt.Errorf("failed to create dns services service to create dns record: %w", err)
+	}
+
+	// Build the new DNS Record options.
+	cnameRecord, err := dnsService.NewResourceRecordInputRdataRdataCnameRecord(cname)
+	if err != nil {
+		return fmt.Errorf("%w", err)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Oh updated with a message.

**Comment by cjschaef** [comment]
`altinfra-e2e-ibmcloud-capi-ovn` failure is expected, ATM, CAPI development is still in progress (specifically LB reconciliation).
```
level=error msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed provisioning resources after infrastructure ready: load balancers missing from ibmcloudCluster.Status in InfraReady
```

**Pull request merged**